Assets/Mirage/Weaver/TypeExtensions.cs

```diff
@@ -98,5 +111,29 @@ public static MethodDefinition AddMethod(this TypeDefinition typeDefinition, str
 
         public static MethodDefinition AddMethod(this TypeDefinition typeDefinition, string name, MethodAttributes attributes) =>
             AddMethod(typeDefinition, name, attributes, typeDefinition.Module.ImportReference(typeof(void)));
+
+        /// <summary>
+        /// Creates a generic type out of another type, if needed.
+        /// </summary>
+        /// <param name="type"></param>
+        /// <returns></returns>
+        public static TypeReference ConvertToGenericIfNeeded(this TypeDefinition type)
+        {
+            if (type.HasGenericParameters)
+            {
+                // get all the generic parameters and make a generic instance out of it
+                TypeReference[] genericTypes = new TypeReference[type.GenericParameters.Count];
+                for (int i = 0; i < type.GenericParameters.Count; i++)
+                {
+                    genericTypes[i] = type.GenericParameters[i].GetElementType();
+                }
+
+                return type.MakeGenericInstanceType(genericTypes);
+            }
```

Comment:
I don't understand what this is doing.  
Let's say you have this type definition:

```
class Myclass<T> {
}
```

then the generic parameter is T,  and making a generic instance with a generic parameter makes no sense.  you want to make a generic instance with a concrete type such as int, or string,  not T.

Are you sure this is needed?

Reply:
Yes, I'm 99% sure this is needed. Looking at compiler-generated code, they always reference their own fields using `MyClass<T>.myField`. This is mainly for fields that are in the generic class.